### PR TITLE
Tls https Client

### PR DESCRIPTION
**Pull request by mgold1234**
add TLS path to configuration,
this gives us the ability to connect in HTTPS with certification from clowder.
related to the story - https://issues.redhat.com/browse/THEEDGE-2972

FIXES: 

## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -147,6 +150,8 @@ func CreateEdgeAPIConfig() (*EdgeConfig, error) {
 	options.SetDefault("KafkaRequestRequiredAcks", -1)
 	options.SetDefault("KafkaMessageSendMaxRetries", 15)
 	options.SetDefault("KafkaRetryBackoffMs", 100)
+	options.SetDefault("TlsCAPath", "")
```
Comment (ldjebran) [inline config/config.go:153]: if there is no value, we can omit  SetDefault as anyway zero value is an empty string

**Review by ldjebran — COMMENTED** [review]
we are creating only the configuration tlscaCert , how this will help us trust the certs.

File: main.go [hunk 2]
```diff
@@ -180,6 +216,7 @@ func main() {
 		// Report captured errors to GlitchTip
 		defer sentry.Recover()
 	}
+	configureHttpClient(*cfg)
```
Comment (ldjebran) [inline main.go:219]: how this call will help us configure our http client image-builder playbook-dipacher ?

File: config/config.go [hunk 3]
```diff
@@ -66,6 +67,8 @@ type EdgeConfig struct {
 	TenantTranslatorURL        string                    `json:"tenant_translator_url,omitempty"`
 	ImageBuilderOrgID          string                    `json:"image_builder_org_id,omitempty"`
 	GlitchtipDsn               string                    `json:"glitchtip_dsn,omitempty"`
+	HTTPClientTimeout          time.Duration             `json:"-"`
```
Comment (AaronH88) [inline config/config.go:70]: needs a correct json value here

File: deploy/clowdapp.yaml [hunk 4]
```diff
@@ -22,6 +22,12 @@ objects:
     name: glitchtip-dsn
   stringData:
     dsn: ${GLITCHTIP_DSN}
+- apiVersion: v1
+  kind: Secret
+  metadata:
+    name: tls-CA-path
+  stringData:
+    dsn: ${TLS_CA_PATH}
```
[the remark this answers is not in the document]
Comment (AaronH88) [reply deploy/clowdapp.yaml:30]: Are you sure dsn is the key?

File: main.go [hunk 5]
```diff
@@ -160,6 +163,39 @@ func featureFlagsServiceUnleash() bool {
...
+		}
+
+		// disable "G402 (CWE-295): TLS MinVersion too low. (Confidence: HIGH, Severity: HIGH)"
+		// #nosec G402
+		httpConfig := &tls.Config{
+			InsecureSkipVerify: false,
+			RootCAs:            rootCAs,
+		}
+		httpTransport := &http.Transport{TLSClientConfig: httpConfig}
+		client := &http.Client{Transport: httpTransport,
```
Comment (AaronH88) [inline main.go:189]: Does a client need to be created per connection / per request / or per pod ? 

Is this thread safe?

File: main.go [hunk 6]
```diff
@@ -160,6 +163,39 @@ func featureFlagsServiceUnleash() bool {
...
 }
 
+func configureHttpClient(cfg config.EdgeConfig) *http.Client {
+	if cfg.TlsCAPath != "" {
+		rootCAs, _ := x509.SystemCertPool()
+		if rootCAs == nil {
+			rootCAs = x509.NewCertPool()
+		}
+
+		certs, err := ioutil.ReadFile(cfg.TlsCAPath)
```
Comment (ldjebran) [inline main.go:173]: ioutil.ReadFile is Deprecated, use os.ReadFile

File: main.go [hunk 7]
```diff
@@ -160,6 +163,39 @@ func featureFlagsServiceUnleash() bool {
...
+		if err != nil {
+			l.LogErrorAndPanic("Failed to append CA to RootCAs", err)
+		}
+
+		if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
+			log.Info("No certs appended, using system certs only")
+		}
+
+		// disable "G402 (CWE-295): TLS MinVersion too low. (Confidence: HIGH, Severity: HIGH)"
```
[the remark this answers is not in the document]
Comment (mgold1234) [reply main.go:182]: it solve this error -G402: TLS MinVersion too low. (gosec)
                httpConfig := &tls.Config{
                        InsecureSkipVerify: false,
                        RootCAs:            rootCAs,
                }
make[1]: *** [golangci-lint-no-fdo] Error 1

File: main.go (hunk 5 above)
Comment (mgold1234) [reply main.go:189]: I will remove that client to image builder client

File: main.go (hunk 6 above)
Comment (mgold1234) [reply main.go:173]: done, thanks

File: config/config.go (hunk 3 above)
Comment (mgold1234) [reply config/config.go]: @AaronH88 , changed it, thank you

File: deploy/clowdapp.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ldjebran) [reply deploy/clowdapp.yaml]: resolve this as the changes was removed

File: config/config.go [hunk 8]
```diff
@@ -148,6 +151,7 @@ func CreateEdgeAPIConfig() (*EdgeConfig, error) {
 	options.SetDefault("KafkaRequestRequiredAcks", -1)
 	options.SetDefault("KafkaMessageSendMaxRetries", 15)
 	options.SetDefault("KafkaRetryBackoffMs", 100)
+	options.SetDefault("HTTPClientTimeout", 10)
```
Comment (ldjebran) [inline config/config.go:154]: Think there is a problem with setting time duration this should be 10 * time.Seconds or something like that.
Masy be set it as type int and later do the conversion to duration

File: config/config.go [hunk 9]
```diff
@@ -197,17 +201,18 @@ func CreateEdgeAPIConfig() (*EdgeConfig, error) {
...
+		WebPort:           options.GetInt("WebPort"),
+		MetricsPort:       options.GetInt("MetricsPort"),
+		MetricsBaseURL:    options.GetString("MetricsBaseURL"),
+		Debug:             options.GetBool("Debug"),
+		LogLevel:          options.GetString("LOG_LEVEL"),
+		BucketName:        options.GetString("EdgeTarballsBucket"),
+		BucketRegion:      options.GetString("BucketRegion"),
+		RepoTempPath:      options.GetString("RepoTempPath"),
+		OpenAPIFilePath:   options.GetString("OpenAPIFilePath"),
+		HTTPClientTimeout: time.Duration(options.GetInt("HTTPClientTimeout")),
```
Comment (ldjebran) [inline config/config.go:215]: same here for duration comment

File: config/config.go [hunk 10]
```diff
@@ -67,6 +68,8 @@ type EdgeConfig struct {
 	ImageBuilderOrgID          string                    `json:"image_builder_org_id,omitempty"`
 	GpgVerify                  string                    `json:"gpg_verify,omitempty"`
 	GlitchtipDsn               string                    `json:"glitchtip_dsn,omitempty"`
+	HTTPClientTimeout          time.Duration             `json:"HTTP_client_timeout,omitempty"`
```
Comment (ldjebran) [inline config/config.go:71]: definitely I see this must be int

File: pkg/clients/client.go [hunk 11]
```diff
@@ -28,3 +33,34 @@ func GetOutgoingHeaders(ctx context.Context) map[string]string {
...
+	if rootCAs == nil {
+		rootCAs = x509.NewCertPool()
+	}
+
+	certs, err := os.ReadFile(cfg.TlsCAPath)
+	if err != nil {
+		l.LogErrorAndPanic("Failed to append CA to RootCAs", err)
+	}
+
+	if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
```
Comment (ldjebran) [inline pkg/clients/client.go:52]: what is the value of cersts if os.ReadFile(cfg.TlsCAPath) fails ?

Comment (ldjebran) [reply pkg/clients/client.go:52]: or should we append certs if ReadFile failed ?

File: config/config.go (hunk 8 above)
Comment (ldjebran) [reply config/config.go:154]: I see in later code you use time.second , from where you got this number why not 30 :) ?

File: pkg/clients/client.go [hunk 12]
```diff
@@ -28,3 +33,34 @@ func GetOutgoingHeaders(ctx context.Context) map[string]string {
 	}
 	return headers
 }
+
+func ConfigureHttpClient(client *http.Client) *http.Client {
+	cfg := config.Get()
+	if cfg.TlsCAPath == "" {
```
Comment (ldjebran) [inline pkg/clients/client.go:39]:
```go
client.Timeout = time.Second * cfg.HTTPClientTimeout
if cfg.TlsCAPath == "" {
    return client
}

```

and remove last last that set client.Timeout

**Review by ldjebran — COMMENTED** [review]
looks good left some comments

File: config/config.go (hunk 8 above)
Comment (mgold1234) [reply config/config.go:154]: I guess I just choose a number, no meaning, do you think 30 is better? I will change it

File: config/config.go (hunk 10 above)
Comment (AaronH88) [reply config/config.go:71]: +1

File: pkg/clients/client.go (hunk 11 above)
Comment (AaronH88) [reply pkg/clients/client.go:53]: as of line 50, we will panic and crash the whole pod, 

Im not sure if that is the best approach

**Comment by AaronH88** [comment]
Getting close,

Can we have some tests to show the imagebuilder client being correctly set up with certs please?

File: pkg/clients/client.go (hunk 11 above)
Comment (ldjebran) [reply pkg/clients/client.go:53]: definitely not the best approach to panic

Comment (mgold1234) [reply pkg/clients/client.go:53]: it's better to put error instead?

File: pkg/clients/client.go (hunk 12 above)
Comment (mgold1234) [reply pkg/clients/client.go:40]: why we need to parse the duration?

File: pkg/clients/client.go [hunk 13]
```diff
@@ -28,3 +33,36 @@ func GetOutgoingHeaders(ctx context.Context) map[string]string {
...
+		log.WithFields(log.Fields{"error": err.Error()}).Error("Failed to parse duration")
+		return client
+	}
+	client.Timeout = timeout
+	if cfg.TlsCAPath == "" {
+		return client
+	}
+	rootCAs, err := x509.SystemCertPool()
+	if err != nil {
+		log.WithFields(log.Fields{"error": err.Error()}).Error("Failed to cert from system cert pool")
```
Comment (ldjebran) [inline pkg/clients/client.go:50]: "failed to get system cert pool"

File: pkg/clients/client.go [hunk 14]
```diff
@@ -28,3 +33,36 @@ func GetOutgoingHeaders(ctx context.Context) map[string]string {
...
+		rootCAs = x509.NewCertPool()
+	}
+	certs, err := os.ReadFile(cfg.TlsCAPath)
+	if err != nil {
+		log.WithFields(log.Fields{"error": err.Error()}).Error("Failed to append CA to RootCAs")
```
Comment (ldjebran) [inline pkg/clients/client.go:55]: "failed to read TlsCaPath"

File: pkg/clients/client.go [hunk 15]
```diff
@@ -28,3 +33,36 @@ func GetOutgoingHeaders(ctx context.Context) map[string]string {
 	}
 	return headers
 }
+
+func ConfigureHttpClient(client *http.Client) *http.Client {
+	cfg := config.Get()
+	timeout, err := time.ParseDuration(fmt.Sprintf("%ds", cfg.HTTPClientTimeout))
+	if err != nil {
+		log.WithFields(log.Fields{"error": err.Error()}).Error("Failed to parse duration")
```
Comment (ldjebran) [inline pkg/clients/client.go:41]: log also the timeout value

File: pkg/clients/client.go [hunk 16]
```diff
@@ -28,3 +33,36 @@ func GetOutgoingHeaders(ctx context.Context) map[string]string {
...
+		log.WithFields(log.Fields{"error": err.Error()}).Error("Failed to cert from system cert pool")
+		rootCAs = x509.NewCertPool()
+	}
+	certs, err := os.ReadFile(cfg.TlsCAPath)
+	if err != nil {
+		log.WithFields(log.Fields{"error": err.Error()}).Error("Failed to append CA to RootCAs")
+		return client
+	}
+	if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
+		log.Warn("adding certificate from PEM failed")
```
Comment (ldjebran) [inline pkg/clients/client.go:59]: log also the TlsCaPath

**Review by ldjebran — COMMENTED** [review]
looks good , left some comments

**Review by ldjebran — APPROVED** [review]
LGTM

**Review by adeliaferreira — APPROVED** [review]
lgtm

**Pull request merged**